### PR TITLE
Proposal: Delete series for object storage

### PR DESCRIPTION
**Pull request by Harshitha1234**
* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

## Changes

Design proposal for delete series #1598

### CODE REVIEW

**Comment by Harshitha1234** [comment]
cc @metalmatze @bwplotka :)

File: pkg/cacheutil/memcached_client.go [hunk 1]
```diff
@@ -269,7 +269,7 @@ func (c *memcachedClient) Stop() {
 	c.workers.Wait()
 }
 
-func (c *memcachedClient) SetAsync(ctx context.Context, key string, value []byte, ttl time.Duration) error {
+func (c *memcachedClient) SetAsync(_ context.Context, key string, value []byte, ttl time.Duration) error {
```
Comment (GiedriusS) [inline pkg/cacheutil/memcached_client.go:272]: Seems like the changes in this file are unrelated? :eye:

File: docs/proposals/deletions_object_storage.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+*   The details are entered via a deletions API (good to have a web UI interface) and they are processed by the compactor to create a tombstone file, if there’s a match for the details entered. Afterwards the tombstone file is uploaded to object storage making it accessible to other components.
+*   If the data with the requested details couldn’t be found in the storage an error message is returned back as a response to the request.
+*   Store Gateway masks the series on processing the tombstones from the object storage.
+*   **Perspectives to deal with Downsampling of blocks having tombstones:**
+    *   **Block with tombstones and max duration to perform deletes passed:** The compactor should first check the maximum duration to perform deletions for that block and if the proposed maxtime passed the deletions are first performed and then the downsampling occurs.
+    *   **Block with tombstones and max duration still hasn’t passed:** Perform compaction.
+    *   **Performing deletes on already compacted blocks:** Have a threshold to perform deletions on the compacted blocks (Prometheus has 5%)
```
Comment (GiedriusS) [inline docs/proposals/deletions_object_storage.md:60]: Could we elaborate more on what is this threshold?

File: docs/proposals/deletions_object_storage.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+*   Tombstones should be append only, so that we can solve these during compaction.
+*   We don’t want to add this feature to the sidecar. The sidecar is expected to be kept lightweight.
+
+
+## Alternatives
+
+
+
+1. External tool operating to perform deletions on the object storage.(Details still to be discussed)
```
Comment (GiedriusS) [inline docs/proposals/deletions_object_storage.md:77]: It's probably unrealistic to have this since the tool would have to coordinate its actions with potentially more than 1 compactor running on a bucket. IMHO it's not worth considering this as a feasible option.

File: docs/proposals/deletions_object_storage.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+*   We propose to implement deletions using the tombstones approach.
+*   A user is expected to enter the following details for performing deletions:
+    *   **external labels**
+    *   **start timestamp**
+    *   **end timestamp** (start and end timestamps of the series data the user expects to be deleted)
+    *   **maximum waiting duration for performing deletions** where a default value is considered if explicitly not specified by the user. (only after this time passes the deletions are performed by the compactor in the next compaction cycle)
+*   The details are entered via a deletions API (good to have a web UI interface) and they are processed by the compactor to create a tombstone file, if there’s a match for the details entered. Afterwards the tombstone file is uploaded to object storage making it accessible to other components.
+*   If the data with the requested details couldn’t be found in the storage an error message is returned back as a response to the request.
+*   Store Gateway masks the series on processing the tombstones from the object storage.
+*   **Perspectives to deal with Downsampling of blocks having tombstones:**
```
Comment (GiedriusS) [inline docs/proposals/deletions_object_storage.md:57]: I guess it's also worth noting that we shouldn't consider blocks for the deletion API operations that have a `deletion-mark.json` i.e. about to be completely deleted themselves.

File: docs/proposals/deletions_object_storage.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    *   **Block with tombstones and max duration to perform deletes passed:** The compactor should first check the maximum duration to perform deletions for that block and if the proposed maxtime passed the deletions are first performed and then the downsampling occurs.
+    *   **Block with tombstones and max duration still hasn’t passed:** Perform compaction.
+    *   **Performing deletes on already compacted blocks:** Have a threshold to perform deletions on the compacted blocks (Prometheus has 5%)
+*   For undoing deletions of a time series there are two proposed ways
+    *   API to undelete a time series - maybe delete the whole tombstones file?
```
Comment (GiedriusS) [inline docs/proposals/deletions_object_storage.md:62]: What if a deletion happens during a compaction pass? In that case, we most likely should return an error to the caller, right?

File: docs/proposals/deletions_object_storage.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    *   “Imaginary” deletion that can delete other tombstones 
+
+Considerations :
+
+
+
+*   Tombstones should be append only, so that we can solve these during compaction.
```
Comment (GiedriusS) [inline docs/proposals/deletions_object_storage.md:69]: `Solve these` ... what is "these"? :P

File: docs/proposals/deletions_object_storage.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+*   We don’t want to add this feature to the sidecar. The sidecar is expected to be kept lightweight.
```
Comment (GiedriusS) [inline docs/proposals/deletions_object_storage.md:70]: :+1: concur with this

File: docs/proposals/deletions_object_storage.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+This design document proposes deletion of series for object storage in Thanos. This feature mainly causes changes in the Compactor and Store components and the object storage itself where the changes are expected to be reflected.
+
+
+# Motivation
+
+The main motivation for considering deletions in the object storage are the following use cases
+
+
+
+*   **Accidental insertion of confidential data:** This is a scenario where a user accidentally inserts confidential data and wants to delete it immediately. In this case the user expects his/her request to accelerate an immediate deletion of the series pertaining to the blocks concerned with that specific data to be deleted.
```
Comment (GiedriusS) [inline docs/proposals/deletions_object_storage.md:28]: We can use gender-neutral language here plus there's a missing comma:

```suggestion
*   **Accidental insertion of confidential data:** This is a scenario where a user accidentally inserts confidential data and wants to delete it immediately. In this case, the user expects their request to accelerate an immediate deletion of the series pertaining to the blocks concerned with that specific data to be deleted.
```

File: docs/proposals/deletions_object_storage.md (hunk 4 above)
Comment (Harshitha1234) [reply docs/proposals/deletions_object_storage.md:57]: Yeah! Definitely that's a good point

File: docs/proposals/deletions_object_storage.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+*   Performing all the above operations at admin level.
+
+
+# Proposed Approach
+
+
+
+*   We propose to implement deletions using the tombstones approach.
+*   A user is expected to enter the following details for performing deletions:
+    *   **external labels**
```
Comment (pracucci) [inline docs/proposals/deletions_object_storage.md:50]: Why external labels and not metric labels?

File: docs/proposals/deletions_object_storage.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    *   **start timestamp**
+    *   **end timestamp** (start and end timestamps of the series data the user expects to be deleted)
+    *   **maximum waiting duration for performing deletions** where a default value is considered if explicitly not specified by the user. (only after this time passes the deletions are performed by the compactor in the next compaction cycle)
+*   The details are entered via a deletions API (good to have a web UI interface) and they are processed by the compactor to create a tombstone file, if there’s a match for the details entered. Afterwards the tombstone file is uploaded to object storage making it accessible to other components.
```
Comment (pracucci) [inline docs/proposals/deletions_object_storage.md:54]: > The details are entered via a deletions API (good to have a web UI interface) and they are processed by the compactor

This means the API is exposed by the compactor. I personally see the compactor more an "off-the-grid" component and I'm not sure that design-wise it's good that it will expose this API.

File: docs/proposals/deletions_object_storage.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+*   If the data with the requested details couldn’t be found in the storage an error message is returned back as a response to the request.
```
Comment (pracucci) [inline docs/proposals/deletions_object_storage.md:55]: > If the data with the requested details couldn’t be found in the storage an error message is returned back as a response to the request.

This means the deletion API request is synchronous. The problem is that on large setups, processing this request may take much time and the request could timeout before done. I think an async approach would be more robust:

1. Deletion API returns a request ID
2. A second API allows to GET the status of a request ID

File: docs/proposals/deletions_object_storage.md [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+1. External tool operating to perform deletions on the object storage.(Details still to be discussed)
+
+
+## Action Plan
+
+
+
+*   Add the deletion API (probably compactor) that only creates tombstones
+*   Store Gateway should be able to mask based on the tombstones from object storage
+*   A web UI for the deletion API?
```
Comment (pracucci) [inline docs/proposals/deletions_object_storage.md:86]: I would keep this proposal smaller and remove the UI part form this proposal.

File: pkg/cacheutil/memcached_client.go (hunk 1 above)
Comment (pracucci) [reply pkg/cacheutil/memcached_client.go:272]: This was done by me in the PR #2775. I think this proposal branch has been created from that PR branch or something like this.

File: docs/proposals/deletions_object_storage.md (hunk 2 above)
Comment (metalmatze) [reply docs/proposals/deletions_object_storage.md:60]: This threshold is something that Prometheus does as well. We need to check for the exact value and maybe should want to link here?

File: docs/proposals/deletions_object_storage.md (hunk 3 above)
Comment (metalmatze) [reply docs/proposals/deletions_object_storage.md:77]: You think we should remove it from the proposal entirely?

File: docs/proposals/deletions_object_storage.md [hunk 13]
```diff
@@ -0,0 +1,86 @@
+# Delete series for object storage
+
+**Date:**	&lt;2020-06-17>
```
Comment (bwplotka) [inline docs/proposals/deletions_object_storage.md:3]: It has to have certain format, please other proposals, something like:

```
---
title: Thanos Remote Write
type: proposal
menu: proposals
status: accepted
owner: brancz
---
```

File: docs/proposals/deletions_object_storage.md (hunk 10 above)
Comment (metalmatze) [reply docs/proposals/deletions_object_storage.md:54]: The compactor is one of the very few components (next to Sidecar, Receiver, Ruler) that has write capabilities to object storage. At the same time it is only ever running once per object storage, so that we have a single writer to the deletion tombstones. 
Additionally, by design, we thought that having the compactor "off-the-grid" and thus only accessible by an administrator solves a few permission problems right away.

We'll add this point to the proposal.

File: docs/proposals/deletions_object_storage.md (hunk 11 above)
Comment (metalmatze) [reply docs/proposals/deletions_object_storage.md:55]: We thought that the creation of tombstones is synchronous and should be able to respond fairly quickly. The deletion itself of times series is still asynchronous and will eventually happen.
The API has an option to list all tombstones and through that get their status.

Because deletion is asynchronous and the creation of tombstones synchronous but fairly quick, we don't think the overhead of a asynchronous approach for the tombstones creation is justified.

File: docs/proposals/deletions_object_storage.md (hunk 2 above)
Comment (metalmatze) [reply docs/proposals/deletions_object_storage.md:60]: For now we want to start with the 5% that Prometheus has. 
We want to follow up on that threshold on IRC/Matrix and ask why it was chosen. In the future, we can always be smarter about this.

File: docs/proposals/deletions_object_storage.md (hunk 5 above)
Comment (metalmatze) [reply docs/proposals/deletions_object_storage.md:62]: The compactor is continuously running compactions and return errors to the caller might result in a lot of errors all the time, which would be annoying.

For now, we are going to check how Prometheus handles this.

File: docs/proposals/deletions_object_storage.md (hunk 6 above)
Comment (metalmatze) [reply docs/proposals/deletions_object_storage.md:69]: tombstones, will update.

File: docs/proposals/deletions_object_storage.md (hunk 12 above)
Comment (metalmatze) [reply docs/proposals/deletions_object_storage.md:86]: Totally, this proposal in itself is quite complex and we will remove the UI from for now. Later, we will follow up on this.

**Comment by Harshitha1234** [comment]
Thank you for the feedback all the points were really worth noting :muscle:

**Comment by Harshitha1234** [comment]
@bwplotka @GiedriusS @metalmatze @pracucci :hugs:

File: docs/proposals/deletions_object_storage.md (hunk 2 above)
Comment (Harshitha1234) [reply docs/proposals/deletions_object_storage.md]: Adding this to the Prometheus community meeting, happening next week :)

File: docs/proposals/deletions_object_storage.md [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+    *   **maximum waiting duration for performing deletions** where a default value is considered if explicitly not specified by the user. (only after this time passes the deletions are performed by the compactor in the next compaction cycle)
+*   The details are entered via a deletions API and they are processed by the compactor to create a tombstone file, if there’s a match for the details entered. Afterwards the tombstone file is uploaded to object storage making it accessible to other components.
+*   **Why Compactor**? : The compactor is one of the very few components (next to Sidecar, Receiver, Ruler) that has write capabilities to object storage. At the same time it is only ever running once per object storage, so that we have a single writer to the deletion tombstones.
+*   If the data with the requested details couldn’t be found in the storage an error message is returned back as a response to the request.
+*   Store Gateway masks the series on processing the tombstones from the object storage.
+*   **Perspectives to deal with Downsampling of blocks having tombstones:**
```
Comment (bwplotka) [inline docs/proposals/deletions_object_storage.md:43]: ` **Perspectives to deal with Downsampling of blocks having tombstones:**` Unfinished?

Comment (bwplotka) [reply docs/proposals/deletions_object_storage.md:43]: Oh you mean `Compactions` maybe?

Comment (bwplotka) [inline docs/proposals/deletions_object_storage.md:43]:
```suggestion
*   **Perspectives to deal with `compaction` of blocks having tombstones:**
```

File: docs/proposals/deletions_object_storage.md [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+    *   **Block with tombstones and max duration to perform deletes passed:** The compactor should first check the maximum duration to perform deletions for that block and if the proposed maxtime passed the deletions are first performed and then the downsampling occurs.
```
Comment (bwplotka) [inline docs/proposals/deletions_object_storage.md:44]:
```suggestion
    *   **Block with tombstones and max duration to perform deletes passed:** The compactor should first check the maximum duration to perform deletions for that block and if the proposed maxtime passed the deletions are first performed and then the compaction occurs.
```

File: docs/proposals/deletions_object_storage.md [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+
+## Proposed Approach
+
+*   We propose to implement deletions using the tombstones approach.
+*   To start off we use the Prometheus tombstone file format.
+*   A user is expected to enter the following details for performing deletions:
+    *   **label matchers**
+    *   **start timestamp**
+    *   **end timestamp** (start and end timestamps of the series data the user expects to be deleted)
+    *   **maximum waiting duration for performing deletions** where a default value is considered if explicitly not specified by the user. (only after this time passes the deletions are performed by the compactor in the next compaction cycle)
```
Comment (bwplotka) [inline docs/proposals/deletions_object_storage.md:38]: I think I am missing what this is, can we reword?

Comment (Harshitha1234) [reply docs/proposals/deletions_object_storage.md]: Moved this feature to the future work :)

File: docs/proposals/deletions_object_storage.md (hunk 14 above)
Comment (Harshitha1234) [reply docs/proposals/deletions_object_storage.md]: My bad! :facepalm:

**Comment by Harshitha1234** [comment]
Made new changes as per the discussions in my weekly meeting with @metalmatze @bwplotka. :) 

TODOs from our meeting:
- Add more info about what we do with compactions
- What to do if compaction level is max and there are tombstones
- Postpone undeletes feature & max waiting period feature considered while performing actual deletions & separate component to later proposal
- Keep design clear what’s now what’s later.
- Downsampling : copy tombstones

File: docs/proposals/deletions_object_storage.md [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+*   **Deletions to sustain user requirements:** Let’s assume the user has some data which leads to some unexpected results or causes performance degradation (due to high cardinality) and the user wants to restore the previous data set-up for obtaining the desired results. In this scenario, the user would want to send a request to mask the data for the time being as there isn’t a high priority requirement to delete the data but eventually during the compaction the user can expect the data to be deleted by the compactor not leading to any major performance issues.
+*   **Helps achieving series based retention (e.g rule aggregation) [#903](https://github.com/thanos-io/thanos/issues/903).**
+
+## Goals
+
+*   Unblock users and allow series deletion in the object storage using tombstones.
+*   Deal with compaction and downsampling of blocks with tombstones.
+*   Performing deletions at admin level.
+
+## Proposed Approach
```
Comment (pracucci) [inline docs/proposals/deletions_object_storage.md:29]: How to deal with series that have been already ingested in the receiver, but not compacted into a block uploaded to the bucket yet?

Comment (yeya24) [reply docs/proposals/deletions_object_storage.md:29]: This proposal is only about deleting series in objstore. So maybe we need to a separate series deletion API for deleting series in TSDB components (receiver, ruler), what do you think?

File: docs/proposals/deletions_object_storage.md [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+*Edge case:* Do we rebuild all the rows of a tombstone? Because we need to be aware that the tombstone is being included(or not) in the given Store API call.
+
+*We're starting with (2) Idea 2*
+
+#### Considerations :
+
+*   Tombstones should be append only, so that we can solve tombstones and rewrite the blocks by performing changes. The old block and the tombstones are deleted during compaction.
+*   We don’t want to add this feature to the sidecar. The sidecar is expected to be kept lightweight.
+
+## Alternatives
```
[the remark this answers is not in the document]
Comment (Harshitha1234) [reply docs/proposals/deletions_object_storage.md:67]: Yeah.. Sure! :)

File: docs/proposals/deletions_object_storage.md [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+*   **Why Compactor**? : The compactor is one of the very few components (next to Sidecar, Receiver, Ruler) that has write capabilities to object storage. At the same time it is only ever running once per object storage, so that we have a single writer to the deletion tombstones.
+*   If the data with the requested details couldn’t be found in the storage an error message is returned back as a response to the request.
+*   Store Gateway masks the series on processing the tombstones from the object storage.
+*   **Perspectives to deal with Compaction of blocks having tombstones:**
+    *   **Block with tombstones** Have a threshold to perform deletions on the compacted blocks ([In Prometheus](https://github.com/prometheus/prometheus/blob/f0a439bfc5d1f49cec113ee9202993be4b002b1b/tsdb/compact.go#L213), the blocks with big enough time range, that have >5% tombstones, are considered for compaction.) We solve the tombstones, if the tombstones are greater than than the threshold and then perform compaction. If not we attach the tombstone file to the new block. If multiple blocks are being compacted, we merge the tombstone files of the blocks whose threshold is not met.
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply docs/proposals/deletions_object_storage.md:42]: Yes. I think the 5% rule should apply only for blocks that have a max level of compaction. Otherwise, all should be applied during compaction as it's essentially almost free to apply deletions during compaction.

The only caveat was about some `delay` time for tombstones to ensure undelete. This functionality was hower decided to be handled in later efforts, right?

File: docs/proposals/deletions_object_storage.md [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+    *   **Block with deletion-mark.json i.e., entire block marked for deletion:** Returns an error message as the entire block is going to be deleted.
+*   **Perspectives to deal with Downsampling of blocks:**
+    *   **Block with tombstones:** If the tombstones are less than the threshold we copy the tombstone file and attach it to the new downsampled block else we solve the tombstones and downsample the block. And the downsampled block with tombstones during its next compaction would again have the same cases as with the compaction of a block with tombstones.
+    *   **Blocks without tombstones:** Downsampling happens...
+
+##### Problems during implementations and possible ideas to solve :
+During the implementation phase of this proposal one of the first problems we came across was that we had to pull the index of all the blocks for creating or appending a tombstone.
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply docs/proposals/deletions_object_storage.md:49]: Yea, Compactor is not responsible for that. It does not load or need to load all the blocks in the system in order to craft tombstones. That's why we cannot use Prometheus format tombstone and have to figure out something else. If we would required to load indexes for ALL blocks in order to create deletion request - this will never scale. 

We need to delegate logic of figuring our which series ID to delete in leafs for querying and during compaction for rewrite.

The tricky part is to know which blocks have the data to be deleted. We thought about simple Series with `skipChunk` options

File: docs/proposals/deletions_object_storage.md (hunk 17 above)
Comment (bwplotka) [reply docs/proposals/deletions_object_storage.md:29]: One thing at the time (: 

I think for this we might need just repeated deletion call - easy if deletion is one of action  (as it should be)

File: docs/proposals/deletions_object_storage.md [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+## Goals
+
+*   Unblock users and allow series deletion in the object storage using tombstones.
+*   Deal with compaction and downsampling of blocks with tombstones.
+*   Performing deletions at admin level.
+
+## Proposed Approach
+
+*   We propose to implement deletions using the tombstones approach.
+*   To start off we use the Prometheus tombstone file format.
```
Comment (bwplotka) [inline docs/proposals/deletions_object_storage.md:32]: Let's make sure to point out why this is not possible, and let's put that to Alternatives

File: docs/proposals/deletions_object_storage.md [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+##### Proposed Alternative Approaches:
+(1) To have a separate store like component with index cache.
+
+(2) To have a different format for the tombstones. Instead of having <seriesRef, min, max> maybe we can have <matcher, min, max>
+* **Idea 1:** Having a CLI tool for dealing with deletions.
+* **Idea 2:** To have all tombstones at some single place and they are used while performing compaction or during query time.
+* **Idea 3:** If we want to have an API for deletions, one way is to have the Store API configured in such a way that the compactor can use it to check for a match.
+*Edge case:* Do we rebuild all the rows of a tombstone? Because we need to be aware that the tombstone is being included(or not) in the given Store API call.
+
+*We're starting with (2) Idea 2*
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply docs/proposals/deletions_object_storage.md:60]: Yes, let's be clear about this.

File: docs/proposals/deletions_object_storage.md [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+    *   **Block with deletion-mark.json i.e., entire block marked for deletion:** Returns an error message as the entire block is going to be deleted.
+*   **Perspectives to deal with Downsampling of blocks:**
+    *   **Block with tombstones:** If the tombstones are less than the threshold we copy the tombstone file and attach it to the new downsampled block else we solve the tombstones and downsample the block. And the downsampled block with tombstones during its next compaction would again have the same cases as with the compaction of a block with tombstones.
+    *   **Blocks without tombstones:** Downsampling happens...
+
+##### Problems during implementations and possible ideas to solve :
+During the implementation phase of this proposal one of the first problems we came across was that we had to pull the index of all the blocks for creating or appending a tombstone.
+
+##### Proposed Alternative Approaches:
```
Comment (bwplotka) [inline docs/proposals/deletions_object_storage.md:51]: Normally we would do 

`## Alternative` sections with couple of alternative, but with clear separation. e.g

```
## Proposed Solution

* CLI.
* Custom format global - single file per request

.. elaborate on how and what challenges / open questions are.
 
## Alternatives

### Where Deletion API sits
 
1. Compactor: 
  * Pros: Easiest as it's single writer in Thanos (but not for Cortex)
  * Cons: Does not have quick access to querying or block indexes
1. Store GW:
  * Pros: Does have quick access to querying or block indexes
  * Cons: There would many writers, and store never was about to write things, it only needed read access

### How we store tombstone and in what format

1. Prometheus Format per block
1. Prometheus Format global
1. Custom format per block
1. Custom format global - appendable file

Comment (bwplotka) [reply docs/proposals/deletions_object_storage.md:51]: We can explain that `* Custom format global - single file per request` idea is good because:
* https://cloud-native.slack.com/archives/CL25937SP/p1595954850430300
* We can easily have multiple writers
* No need to have index data involved

## Chalenges
* How to "compact" / "remove" those files when applied.
* Do we want those deletion to by applied only during compaction or also for already compacted blocks. If yes how? 5%? And how to tell that when is 5%? What component would do that?
* Any rate limiting, what if there are too many files? Would that scale?

Comment (bwplotka) [reply docs/proposals/deletions_object_storage.md:51]: Plus other ideas to talk about later like:

Optimization for file names?

File: docs/proposals/deletions_object_storage.md [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+
+(2) To have a different format for the tombstones. Instead of having <seriesRef, min, max> maybe we can have <matcher, min, max>
+* **Idea 1:** Having a CLI tool for dealing with deletions.
+* **Idea 2:** To have all tombstones at some single place and they are used while performing compaction or during query time.
+* **Idea 3:** If we want to have an API for deletions, one way is to have the Store API configured in such a way that the compactor can use it to check for a match.
+*Edge case:* Do we rebuild all the rows of a tombstone? Because we need to be aware that the tombstone is being included(or not) in the given Store API call.
+
+*We're starting with (2) Idea 2*
+
+#### Considerations :
```
[the remark this answers is not in the document]
Comment (Harshitha1234) [reply docs/proposals/deletions_object_storage.md:62]: It was intended for the original proposal :)

**Comment by bwplotka** [comment]
Hey, can we finalize this and make sure it's clean for other ppl? We just share it to Cortex devs, they would be happy to contribute (:

**Comment by Harshitha1234** [comment]
Yes, sure. PTAL I've made the latest changes based on our last discussion about the proposal, if something doesn't seem right please let me know 🙂

File: docs/proposals/deletions_object_storage.md [hunk 25]
```diff
@@ -0,0 +1,100 @@
...
+*   During compaction or downsampling, we check if the blocks being considered have series corresponding to a tombstone file, if so we delete the data and continue with the process.
+
+## Open Questions
+*   Optimization for file names?
+*   Do we create a tombstone irrespective of the presence of the series with the given matcher? Or should we check for the existence of series that corresponds to the deletion request?
+*   Should we consider having a threshold for compaction or downsampling where we check all the tombstones and calculate the percentage of deletions?
+
+## Considerations
+
+*   The old blocks and the tombstones are deleted during compaction time.
```
Comment (pracucci) [inline docs/proposals/deletions_object_storage.md:52]: I think this opens to a race condition.

Above it's mentioned this:
> Store Gateway masks the series on processing the global tombstone files from the object storage.

If the store gateway scans deleted tombstones before the old blocks are effectively offloaded by the store gateway (we do soft delete first, apply a delay and then delete hard delete them) we may open to a time window during which the tombstone is deleted, the store gateway removed it from its local map, the old block is still queried and we do re-expose deleted series.

File: docs/proposals/deletions_object_storage.md [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
+## Goals
+
+*   Unblock users and allow series deletion in the object storage using tombstones.
+*   Deal with compaction and downsampling of blocks with tombstones.
+*   Performing deletions at admin level.
+
+## Proposed Approach
+
+*   We propose to implement deletions via the tombstones approach using a CLI tool.
+*   A tombstone is proposed to be a **Custom format global - single file per request**.
```
Comment (pracucci) [inline docs/proposals/deletions_object_storage.md:32]: I haven't found the exact format of this file documented. Anyway, I would add the timestamp of the deletion request inside as well. Would allow to build hard deletion grace periods, without having to read it from the object storage (object creation).

**Review by metalmatze — APPROVED** [review]
Would be nice to finally merge. Seems like we at least agree on the first steps so far :blush:

File: docs/proposals/deletions_object_storage.md (hunk 26 above)
Comment (bwplotka) [reply docs/proposals/deletions_object_storage.md:32]: still missing @Harshitha1234

File: docs/proposals/deletions_object_storage.md [hunk 27]
```diff
@@ -0,0 +1,100 @@
...
+*   **Why Custom format global - single file per request**? :
+    *   https://cloud-native.slack.com/archives/CL25937SP/p1595954850430300
```
Comment (bwplotka) [inline docs/proposals/deletions_object_storage.md:34]: ? 

We need more details, can we use full sentence?

Comment (bwplotka) [reply docs/proposals/deletions_object_storage.md:34]: Also design is to make sure we have full view, so I would rather explain instead of linking some further reading (:

File: docs/proposals/deletions_object_storage.md [hunk 28]
```diff
@@ -0,0 +1,100 @@
...
+    *   We can easily have multiple writers
+    *   No need to have index data of the blocks
+*   A user is expected to enter the following details for performing deletions:
+    *   **label matchers**
+    *   **start timestamp**
+    *   **end timestamp** (start and end timestamps of the series data the user expects to be deleted)
```
Comment (bwplotka) [inline docs/proposals/deletions_object_storage.md:40]: What about things like `reason` or `author`? (:

File: docs/proposals/deletions_object_storage.md [hunk 29]
```diff
@@ -0,0 +1,100 @@
...
+*   The entered details are processed by the CLI tool to create a tombstone file (unique for a request), and then the tombstone file is uploaded to the object storage making it accessible to all components.
+*   Store Gateway masks the series on processing the global tombstone files from the object storage.
+*   During compaction or downsampling, we check if the blocks being considered have series corresponding to a tombstone file, if so we delete the data and continue with the process.
+
+## Open Questions
+*   Optimization for file names?
```
Comment (bwplotka) [inline docs/proposals/deletions_object_storage.md:46]: Can you describe more why we would need this? (:

File: docs/proposals/deletions_object_storage.md [hunk 30]
```diff
@@ -0,0 +1,100 @@
...
+*   Do we create a tombstone irrespective of the presence of the series with the given matcher? Or should we check for the existence of series that corresponds to the deletion request?
```
Comment (bwplotka) [inline docs/proposals/deletions_object_storage.md:47]: I think this should be moved to proposal - we would love to avoid doing this.

Comment (bwplotka) [reply docs/proposals/deletions_object_storage.md:47]: We could have this in alternatives

File: docs/proposals/deletions_object_storage.md [hunk 31]
```diff
@@ -0,0 +1,100 @@
...
+*   Should we consider having a threshold for compaction or downsampling where we check all the tombstones and calculate the percentage of deletions?
```
Comment (bwplotka) [inline docs/proposals/deletions_object_storage.md:48]: Yes - this should be covered as well here :thinking: 

As per - how to delete stuff from already compacted blocks section

File: docs/proposals/deletions_object_storage.md (hunk 25 above)
Comment (bwplotka) [reply docs/proposals/deletions_object_storage.md:52]: Yes, it has to be delayed-deleted or something else... It's not easy, because:

* How we know there are no more stuff to delete then in mentioned block?
* What if someone backfills block that touches deleted period?
* What if min-time is .. not specified? Does it has to be?


Behaviour on all those questions have to be defined in this proposal, explicit and simple to understand for user of deletion API I think ):

**Comment by bwplotka** [comment]
Let's get back to this... is it ready for review? :hugs:

**Comment by Harshitha1234** [comment]
Yes, it is :)

**Comment by Harshitha1234** [comment]
Still valid! @bwplotka Are there any key changes that we expect in this proposal? :)

**Review by bwplotka — APPROVED** [review]
No, just implement this 🤗

**Comment by bwplotka** [comment]
Merged! Thanks @Harshitha1234 🤗

**Comment by Harshitha1234** [comment]
> No, just implement this 🤗

Oh yeah! Let's do this then!! 💪

**Comment by yeya24** [comment]
Tombstone files are global. Then is there a way to achieve some multi-tenancy or fine-grained control in this case?
What if I only want to delete series `{xxx}` in blocks with external labels `cluster=eu`? Or only want to delete series in raw blocks but not in downsampled blocks?

These cases cannot be easily handled with global tombstone files.
